// banyan-webcrypto-experiment/lib/types.js
export var ExportKeyFormat;
(function (ExportKeyFormat) {
    ExportKeyFormat["PKCS8"] = "pkcs8";
    ExportKeyFormat["SPKI"] = "spki";
    ExportKeyFormat["RAW"] = "raw";
})(ExportKeyFormat || (ExportKeyFormat = {}));
export var CryptoSystem;
(function (CryptoSystem) {
    CryptoSystem["ECC"] = "ecc";
})(CryptoSystem || (CryptoSystem = {}));
export var EccCurve;
(function (EccCurve) {
    EccCurve["P_384"] = "P-384";
})(EccCurve || (EccCurve = {}));
export var SymmAlg;
(function (SymmAlg) {
    SymmAlg["AES_GCM"] = "AES-GCM";
    SymmAlg["AES_KW"] = "AES-KW";
})(SymmAlg || (SymmAlg = {}));
export var SymmKeyLength;
(function (SymmKeyLength) {
    SymmKeyLength[SymmKeyLength["B256"] = 256] = "B256";
    SymmKeyLength[SymmKeyLength["B384"] = 384] = "B384";
    SymmKeyLength[SymmKeyLength["B512"] = 512] = "B512";
})(SymmKeyLength || (SymmKeyLength = {}));
export var HashAlg;
(function (HashAlg) {
    HashAlg["SHA_256"] = "SHA-256";
    HashAlg["SHA_384"] = "SHA-384";
    HashAlg["SHA_512"] = "SHA-512";
})(HashAlg || (HashAlg = {}));
export var CharSize;
(function (CharSize) {
    CharSize[CharSize["B8"] = 8] = "B8";
    CharSize[CharSize["B16"] = 16] = "B16";
})(CharSize || (CharSize = {}));
export var KeyUse;
(function (KeyUse) {
    KeyUse["Exchange"] = "exchange";
    KeyUse["Write"] = "write";
})(KeyUse || (KeyUse = {}));
